docs/markdown/using-snowchain/validating-and-relaying/relayer-flow.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: default
+title: Relayer Flow
+parent: Validating and Relaying
+nav_order: 2
+---
+
+# Relayer Flow
+
+## Polkadot to Ethereum
+The relayer can be started in 2 different modes for relaying from Polkadot to Ethereum.
+
+## MMR-Epochs-Only mode
+This mode runs the relayer in a mode where it ensures that the Ethereum contract receives at least 1 MMR proof per epoch. It only interacts with the Relay Chain Light Client, not the Parachain Light Client.
+
+### Relayer process starts in MMR-Epochs-Only mode
+```mermaid!
+graph TD;
+  S1[Query Relay Chain Light Client for latest relayed MMR Epoch];
+  S1-->S2;
+  S2[Save as latestRelayedEpoch];
+  S2-->S3;
+  S2-->S5;
+  S3[Start querying/watching ethereum for new blocks];
+  S3-- New Ethereum Block Found -->S4;
+  S4[Add to Ethereum Block Processing Queue];
+  S5[Start querying/watching relay chain for new MMR Roots];
+  S5-- New Relay Chain MMR Root Found -->S6;
+  S6[Add to New MMR Root Processing Queue];
+```
+### Ethereum Block Processing Queue
+```mermaid!
+graph TD;
+  Eb1[New Block In Queue];
+  Eb1-- contains updated latestRelayedEpoch -->Eb2
+  Eb2[Save as latestRelayedEpoch];
+  Eb1-- no update -->Eb3
+  Eb3[Ignore];
+```
+### New MMR Root Processing Queue
+```mermaid!
+graph TD;
+  M1[New MMR Root In Queue];
+  M1-->M1a;
+  M1a[Is epoch == latestRelayedEpoch + 1?];
+  M1a-- No -->M2;
+  M1a-- Yes -->M3;
+  M2[Ignore];
+  M3[Is last block of epoch?];
+  M3-- Yes -->M4;
+  M4[Add new MMR Root to MMR Root Relay queue];
+  M3-- No -->M2;
+```
+### MMR Relay Queue
+```mermaid!
+graph TD;
+  Mr00[New MMR to Relay];
+  Mr00-->Mr01;
+  Mr01[Query relay chain for all signatures on that MMR Root];
+  Mr01-->Mr02;
+  Mr02[Generate Full Merkle Tree of Public Keys and transaction input data];
+  Mr02-->Mr0;
+  Mr0[Submit initial verification transaction to relay chain light client];
+  Mr0--Transaction Successful -->Mr1;
+  Mr1[Watch and wait for x Ethereum blocks until block wait period is over];
+  Mr1-->Mr10;
+  Mr10[Query blockhash for random seed];
+  Mr10-->Mr11;
+  Mr11[Generate random number, selected Merkle Proofs and transaction input data];
+  Mr11-->Mr2;
+  Mr2[Submit final verification transaction to relay chain light client];
+  Mr2--Transaction Successful -->Mr3;
+  Mr3[Done, Ethereum Block Processing Queue should incidentally receive this new block and update latestRelayedEpoch];
+```
```

Comment:
I thought a bit more about the mechanism for preventing multiple competing relayers starting the process for the same block height.

It should be enough to just prevent completion for any relay chain block height that already has a completed MMR root. That just requires one additional storage entry (maybe `latestCompletedHeight`). 

With that approach, multiple completions will never occur, just multiple initiations, which are not that expensive.

Reply:
yep this works unless multiple relayers submit their completion transaction in the same ethereum block, which is pretty much guaranteed in a hypercompetitive environment?

Reply:
Not sure that is correct, the second transaction from the same block would still be blocked? The only exception would be a re-entrancy problem, but there is no call to external contracts.

Reply:
yeh the second transaction in the block will fail and revert, but the relayer then still ends up having to pay gas costs for the calldata, which is likely to be quite expensive still, so ideally still to be avoided in future optimization